### DISCUSSION TITLE
Feature Request: struct aggregate

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#10153] [Ideas]
I'd like to turn the "unnested" table 
```sql
FROM (VALUES ('Alice', 'Left', 4), ('Alice', 'Right', 5), ('Bob', 'Left', 5), ('Bob', 'Right', 5)) longfingers(name, side, number);
```
```
┌─────────┬─────────┬────────┐
│  name   │  side   │ number │
│ varchar │ varchar │ int32  │
├─────────┼─────────┼────────┤
│ Alice   │ Left    │      4 │
│ Alice   │ Right   │      5 │
│ Bob     │ Left    │      5 │
│ Bob     │ Right   │      5 │
└─────────┴─────────┴────────┘
```
into the nested table 
```sql
FROM (VALUES ('Alice', {'Left': 4, 'Right': 5}), ('Bob', {'Left': 5, 'Right': 5})) fingers(name, numbers);
```
```
┌─────────┬─────────────────────────────────────────┐
│  name   │                 numbers                 │
│ varchar │ struct("left" integer, "right" integer) │
├─────────┼─────────────────────────────────────────┤
│ Alice   │ {'Left': 4, 'Right': 5}                 │
│ Bob     │ {'Left': 5, 'Right': 5}                 │
└─────────┴─────────────────────────────────────────┘
```

I was thinking 
```sql
SELECT name, struct(side, number) AS numbers FROM longfingers GROUP BY name
```
should just work, given that this seems to be the main purpose of structs according to the documentation:

```
STRUCTs are typically used to nest multiple columns into a single column
```

The very similar 
```sql
SELECT name, map(list(side), list(number)) AS numbers FROM longfingers GROUP BY name
```
does work, but then I found no way of casting the map to a struct (neither `::STRUCT`, nor `CAST(... AS STRUCT)` worked).
Also, I don't understand why the `list`s inside the map are necessary.

**Comment by soerenwolfers** [comment]
I'd be happy to implement this myself, once someone confirms what the syntax should be.

**Reply by Maxxen** [reply]
Hey! Just trying to understand how this would work, what happens if you have e.g.

```
┌─────────┬─────────┬────────┐
│  name   │  side   │ number │
│ varchar │ varchar │ int32  │
├─────────┼─────────┼────────┤
│ Alice   │ Left    │      4 │
│ Alice   │ Left   │      6 │
│ Bob     │ Left    │      5 │
│ Bob     │ Right   │      5 │
│ Bob     │ Top   │      1337 │
└─────────┴─────────┴────────┘
```

This isn't really nesting multiple _columns_ into a single column, but instead collapsing multiple _rows_ into a single column. If that's what you want to do I think it makes mores sense to use the `list` aggregate instead? As there's no guarantee that each group will contain the same amount of "rows", what fields will the type of the resulting struct contain?

**Reply by soerenwolfers** [reply]
Okay, I thought "nesting" in the docs meant the opposite of "unnest", which does increase number of rows. I do now also see how the dependence of the output type on the row data makes this impossible in SQL.
